File: src/basset_predict.lua
#!/usr/bin/env th

require 'hdf5'

require 'postprocess'

----------------------------------------------------------------
-- parse arguments
----------------------------------------------------------------
cmd = torch.CmdLine()
cmd:text()
cmd:text('DNA ConvNet testing')
cmd:text()
cmd:text('Arguments')
cmd:argument('model_file')
cmd:argument('data_file')
cmd:argument('out_file')
cmd:text()
cmd:text('Options:')
cmd:option('-batch', 128, 'Batch size')
cmd:option('-cuda', false, 'Run on GPGPU')
cmd:option('-cudnn', false, 'Run on GPGPU w/ cuDNN')
cmd:option('-mc_n', 0, 'Perform MCMC prediction')
cmd:option('-norm', false, 'Normalize all targets to a level plane')
cmd:option('-rc', false, 'Average forward and reverse complement')
cmd:option('-scores', false, 'Print scores instead of predictions')
cmd:text()
opt = cmd:parse(arg)

-- set cpu/gpu
cuda_nn = opt.cudnn
cuda = opt.cuda or opt.cudnn
require 'convnet'

----------------------------------------------------------------
-- load data
----------------------------------------------------------------
local data_open = hdf5.open(opt.data_file, 'r')
local test_seqs = data_open:read('test_in')

----------------------------------------------------------------
-- construct model
----------------------------------------------------------------
-- initialize
local convnet = ConvNet:__init()

-- load from saved parameters
local convnet_params = torch.load(opt.model_file)
convnet:load(convnet_params)

----------------------------------------------------------------
-- predict and test
----------------------------------------------------------------
local preds, scores
if opt.mc_n > 1 then
    -- set stochastic evaulate mode
    convnet:evaluate_mc()

    -- compute predictions
    preds, scores = convnet:predict_mc(test_seqs, opt.mc_n, opt.batch, false, opt.rc)
else
    -- set evaluate mode
    convnet.model:evaluate()

    -- compute predictions
    preds, scores = convnet:predict(test_seqs, opt.batch, false, opt.rc)
end

if opt.norm then
    preds = troy_norm(preds, convnet.pred_means)
end

-- close HDF5
data_open:close()

----------------------------------------------------------------
-- dump to file
----------------------------------------------------------------
local values = preds
if opt.scores then
    values = scores
end

local predict_out = io.open(opt.out_file, 'w')

-- print predictions
for si=1,(#values)[1] do
    predict_out:write(values[{si,1}])
    for ti=2,(#values)[2] do
        predict_out:write(string.format("\t%s",values[{si,ti}]))
    end
    predict_out:write("\n")
end

predict_out:close()

-- local hdf_out = hdf5.open(opt.out_file, 'w')
-- hdf_out:write("preds", preds)
-- hdf_out:close()

File: src/basset_predict.py
#!/usr/bin/env python
from optparse import OptionParser
import os, subprocess, tempfile

import h5py

import dna_io

'''
basset_predict.py

Predict a set of test sequences.
'''

################################################################################
# main
############################s####################################################
def main():
    usage = 'usage: %prog [options] <model_file> <input_file> <output_file>'
    parser = OptionParser(usage)
    parser.add_option('-b', '--batch', dest='batch', default=128, type='int', help='Batch size [Default: %default]')
    parser.add_option('--cuda', dest='cuda', default=False, action='store_true', help='Run on GPGPU [Default: %default]')
    parser.add_option('--cudnn', dest='cudnn', default=False, action='store_true', help='Run on GPGPU w/cuDNN [Default: %default]')
    parser.add_option('-n', '--norm', dest='norm', default=False, action='store_true', help='Normalize all targets to a level plane [Default: %default]')
    parser.add_option('-r', '--rc', dest='rc', default=False, action='store_true', help='Average forward and reverse complement [Default: %default]')
    parser.add_option('-s', '--scores', dest='scores', default=False, action='store_true', help='Print pre-sigmoid scores instead of probability predictions [Default: %default]')
    (options,args) = parser.parse_args()

    if len(args) != 3:
        parser.error('Must provide Basset model file, input sequences (as a FASTA file or test data in an HDF file, and output table file')
    else:
        model_file = args[0]
        input_file = args[1]
        out_file = args[2]

    #################################################################
    # parse input file
    #################################################################
    try:
        # input_file is FASTA

        # load sequences
        seqs_1hot = dna_io.load_sequences(input_file, permute=False)

        # reshape sequences for torch
        seqs_1hot = seqs_1hot.reshape((seqs_1hot.shape[0],4,1,seqs_1hot.shape[1]/4))

        # write as test data to a HDF5 file
        model_input_fd, model_input_hdf5 = tempfile.mkstemp()
        h5f = h5py.File(model_input_hdf5, 'w')
        h5f.create_dataset('test_in', data=seqs_1hot)
        h5f.close()
        temp_hdf5 = True

    except (IOError, IndexError):
        # input_file is HDF5
        model_input_hdf5 = input_file
        temp_hdf5 = False


    #################################################################
    # Torch predict modifications
    #################################################################
    opts_str = '-batch %d' % options.batch
    if options.cudnn:
        opts_str += ' -cudnn'
    elif options.cuda:
        opts_str += ' -cuda'
    if options.norm:
        opts_str += ' -norm'
    if options.rc:
        opts_str += ' -rc'
    if options.scores:
        opts_str += ' -scores'

    torch_cmd = '%s/src/basset_predict.lua %s %s %s %s' % (os.environ['BASSETDIR'],opts_str, model_file, model_input_hdf5, out_file)
    print torch_cmd
    subprocess.call(torch_cmd, shell=True)

    if temp_hdf5:
        os.remove(model_input_hdf5)


################################################################################
# __main__
################################################################################
if __name__ == '__main__':
    main()
    #pdb.runcall(main)
